content/blog/writing-better-angular-tests/index.md

```diff
@@ -0,0 +1,208 @@
+
+```js
+it('should create the app', () => {
+  const fixture = TestBed.createComponent(AppComponent);
+  const app = fixture.componentInstance;
+  expect(app).toBeTruthy();
+});
+```
+
+This test isn't a very good test. It doesn't say anything about the actual output of the application component itself. When the output is a full rich webpage and tests are testing Angular then the tests won't do much when the content of the webpage is changed. 
+
+While the default testing setup does allow for the writing of tests that would test the outputted HTML they are still specific to Angular
+
+```js
+it('should render title', () => {
+  const fixture = TestBed.createComponent(AppComponent);
+  fixture.detectChanges();
+  const compiled = fixture.nativeElement;
+  expect(compiled.querySelector('.content span').textContent).toContain('The app is running!');
+});
+```
+
+That test looks a little better but it's still very tied to Angular, require in depth knowledge of how Angular actually routes and moves all the bits around to write tests for it and now the tests are completely tied into Angular and the current API footprint. If over the years Angular is retired these tests are now no longer valid.
+
+If the tests were just tailored to the outputted DOM or containers it would be a much easier and more adaptable test.
+
+```js
+test('should render counter', async () => {
+  await render(AppComponent);
+  expect(document.querySelector('.content span').innerText).toBe('The app is running!');
+});
+```
+
+This test no longer even needs Angular to be the library chosen. It just requires that a render method, when given the component, will render it to the DOM present in the testing environment. This can be run in the Framework, and even tested against in a real world browser. This is a good test in that the first `span` inside of `.content` has the `innerText` value expected in the test. These are all JavaScript and DOM APIs and thus can be trusted in any environment that adheres to them.
+
+Writing tests like this that don't rely on testing Angular but instead rely on on the DOM allows the application to be tested in a way that a user would use the application instead of the way that Angular internally works to display the application.
+
+# Fixing that shortcoming using Testing Library
+
+Thankfully writing tests like these have been made simple by a testing library simply called [Testing Library](https://testing-library.com) and it has an [Angular Testing Library](https://testing-library.com/docs/angular-testing-library/intro) that can be used to test Angular apps in the simple DOM focused manner with some nice helpers to make it even easier to work with. It relies on [Jest](https://jestjs.io/) as an extension to the Jasmine testing framework to make testing easier and more end results focused. With that tooling a project can have tests much less focused on Angular and much more focused on what is being made.

```

Comment:
```suggestion
Thankfully, writing tests like these have been made simple by a testing library simply called "[Testing Library](https://testing-library.com)." Testing Library is a collection of libraries for various frameworks and applications. One of the supported libraries is Angular, through the [Angular Testing Library](https://testing-library.com/docs/angular-testing-library/intro). This can be used to test Angular apps in a simple DOM focused manner with some nice helpers to make it even easier to work with. It relies on [Jest](https://jestjs.io/) as an extension to the Jasmine testing framework to make testing easier, and more end-results focused. With that tooling, a project can have tests much less focused on Angular and much more focused on what is being made.
```